OnPod/Node/Examples/Solenoids/solenoid.cpp

```diff
@@ -0,0 +1,48 @@
+#include <Arduino.h>
+#include "solenoid.h"
+
+DRV8806::DRV8806(uint8_t PIN_LATCH, uint8_t PIN_DOUT, uint8_t PIN_DIN, uint8_t PIN_SCLK) :
+    PIN_LATCH(PIN_LATCH), PIN_DOUT(PIN_DOUT),
+    PIN_DIN(PIN_DIN), PIN_SCLK(PIN_SCLK)
+    {}
+
+void DRV8806::init() {
+    pinMode(PIN_SCLK, OUTPUT);
+    pinMode(PIN_LATCH, OUTPUT);
+    pinMode(PIN_DOUT, INPUT); // DRV8806 => SAMD21
+    pinMode(PIN_DIN, OUTPUT); // SAMD21 => DRV8806
+    digitalWrite(PIN_SCLK, LOW);
+    digitalWrite(PIN_LATCH, HIGH);
+    digitalWrite(PIN_DOUT, LOW);
+}
+
+void DRV8806::updateSolenoids() {
+    digitalWrite(PIN_LATCH, LOW);
+    digitalWrite(PIN_SCLK, LOW);
+    delayMicroseconds(10);
+
+    // loop through iActiveSolenoids and write DIN to the value of each bit
+    // then write SCLK high then low to shift the bit in
+    for (uint8_t i = 0; i < 8; i++) {
+        uint8_t iSolenoid = 7 - i;
+        bool bState = (iActiveSolenoids >> iSolenoid) & 1;
+        digitalWrite(PIN_DIN, bState);
+        delayMicroseconds(1);
+        digitalWrite(PIN_SCLK, HIGH);
+        delayMicroseconds(1);
+        digitalWrite(PIN_SCLK, LOW);
+        delayMicroseconds(1);
+    }
+    delayMicroseconds(10);
+    digitalWrite(PIN_LATCH, HIGH);
+}
+
+void DRV8806::enableSolenoid(uint8_t iSolenoid) {
+    // set the selected iActiveSolenoids bit to 1
+    iActiveSolenoids |= (1 << iSolenoid);
+}
+
+void DRV8806::disableSolenoid(uint8_t iSolenoid) {
+    // set the selected iActiveSolenoids bit to 0
+    iActiveSolenoids &= ~(1 << iSolenoid);
```

Comment:
iSolenoid can be much greater then the 8 bits of iActiveSolenoids so ive seen this commonly wrapped with a 0xFF mask. The behaviour should be the same without the mask here, but something to keep in mind with variables on the right side of shifting.

Reply:
I understand what you mean about iSolenoids being able to be larger than the number of bits in iActiveSolenoids but what do you mean by "wrapping" it in 0xFF?

Reply:
(~(1 << iSolenoid) & 0xFF)

Reply:
then your just left with 8 bits, but in this case its fine since the overflow should be truncated off